### PR TITLE
chore(cuda): add asserts on base log, poly size and num samples values

### PR DESCRIPTION
**Pull request by agnesLeroy**
### Checklist 

(Use '[x]' to check the checkboxes, or submit the PR and then click the checkboxes)

* [x] Tests for the changes have been added (for bug fixes / features)
* [x] Docs have been added / updated (for bug fixes / features)
* [x] The PR description links to the related issue (to link an issue, use '#XXX'.)
* [x] The tests on AWS have been launched and are successful (comment with @slab-ci cpu_test and/or @slab-ci gpu_test to trigger the tests)
* [x] The draft release description has been updated
* [x] Check for breaking changes (including serialization changes) and add them to commit message following the conventional commit [specification][conventional-breaking]

[conventional-breaking]: https://www.conventionalcommits.org/en/v1.0.0/#commit-message-with-description-and-breaking-change-footer

### CODE REVIEW

**Comment by agnesLeroy** [comment]
@bbarbakadze @pdroalves I have added some asserts but the query to get the SM number may be long. Before merging this I think it would be good that I run benchmarks just to check the effect wdyt?

**Comment by agnesLeroy** [comment]
@slab-ci gpu_test

**Comment by pdroalves** [comment]
Hi @agnesLeroy !
I have run benchmakrs comparing the last two commits in this PR (419dd8ee16173c5ed75fdaab1b2709e65289fd05 and b6a18b4be4f444b622bdce5c1bfde90414eb877a) and I couldn't measure any non-negligible difference in performance for low latency and amortized PBSs.


```
pbs/gpu_amortized_pbs_64/num_c=1, n=691, k=1, N=1024, ell=3, bl=6                        
                        time:   [88.037 ms 89.335 ms 90.688 ms]
                        change: [-1.9084% +0.1344% +2.1879%] (p = 0.89 > 0.10)

pbs/gpu_amortized_pbs_64/num_c=10, n=691, k=1, N=1024, ell=3, bl=6                        
                        time:   [91.350 ms 91.780 ms 92.198 ms]
                        change: [-0.2970% +0.2718% +0.8285%] (p = 0.37 > 0.10)
                        No change in performance detected.

pbs/gpu_amortized_pbs_32/num_c=1, n=691, k=1, N=1024, ell=3, bl=6                                                                           
                        time:   [88.271 ms 88.542 ms 88.863 ms]
                        change: [-0.8889% -0.4772% -0.0326%] (p = 0.06 < 0.10)
                        Change within noise threshold.
pbs/gpu_amortized_pbs_32/num_c=10, n=691, k=1, N=1024, ell=3, bl=6                        
                        time:   [88.601 ms 89.287 ms 89.949 ms]
                        change: [-0.8386% +0.1381% +1.0954%] (p = 0.79 > 0.10)
                        No change in performance detected.

pbs/gpu_lowlat_pbs_64/num_c=1, n=691, k=1, N=1024, ell=3, bl=6                                                                           
                        time:   [69.234 ms 69.347 ms 69.473 ms]
                        change: [-1.1383% -0.7714% -0.3582%] (p = 0.00 < 0.10)
                        Change within noise threshold.
pbs/gpu_lowlat_pbs_64/num_c=10, n=691, k=1, N=1024, ell=3, bl=6                                                                           
                        time:   [69.784 ms 70.152 ms 70.488 ms]
                        change: [-0.7846% -0.0180% +0.7824%] (p = 0.97 > 0.10)
                        No change in performance detected.

pbs/gpu_lowlat_pbs_32/num_c=1, n=691, k=1, N=1024, ell=3, bl=6                                                                           
                        time:   [73.978 ms 74.215 ms 74.462 ms]
                        change: [+0.2519% +0.6106% +0.9658%] (p = 0.00 < 0.10)
                        Change within noise threshold.
pbs/gpu_lowlat_pbs_32/num_c=10, n=691, k=1, N=1024, ell=3, bl=6                                                                           
                        time:   [73.993 ms 74.156 ms 74.358 ms]
                        change: [-0.3838% +0.0746% +0.5337%] (p = 0.77 > 0.10)
                        No change in performance detected.
```

**Review by pdroalves — APPROVED** [review]

File: concrete-cuda/cuda/src/bootstrap_amortized.cu [hunk 1]
```diff
@@ -111,7 +111,9 @@ void cuda_bootstrap_amortized_lwe_ciphertext_vector_32(
         num_lut_vectors, lwe_idx, max_shared_memory);
     break;
   default:
-    break;
+    printf("Error: polynomial size should be one of 512, 1024, 2048, 4096, "
```
Comment (mayeul-zama) [inline concrete-cuda/cuda/src/bootstrap_amortized.cu:114]: Some context could be added:
`Gpu Amortized Pbs: Error: polynomial size should be one of 512, 1024, 2048, 4096` for example

In case of the error being seen by a high level user, it helps better understand the cause

File: concrete-cuda/cuda/src/bootstrap_amortized.cu [hunk 2]
```diff
@@ -169,6 +171,8 @@ void cuda_bootstrap_amortized_lwe_ciphertext_vector_64(
...
```
Comment (mayeul-zama) [inline concrete-cuda/cuda/src/bootstrap_amortized.cu:174]: Some context could be added

File: concrete-cuda/cuda/src/bootstrap_amortized.cuh [hunk 3]
```diff
@@ -320,6 +320,8 @@ __host__ void host_bootstrap_amortized(
     uint32_t lwe_idx,
     uint32_t max_shared_memory) {
 
+  assert(base_log <= 16);
```
Comment (mayeul-zama) [inline concrete-cuda/cuda/src/bootstrap_amortized.cuh:323]: A print with some context could be added before aborting

File: concrete-cuda/cuda/src/bootstrap_low_latency.cu [hunk 4]
```diff
@@ -97,24 +97,9 @@ void cuda_bootstrap_low_latency_lwe_ciphertext_vector_32(
...
-    host_bootstrap_low_latency<uint32_t, Degree<8192>>(
-        v_stream, (uint32_t *)lwe_out, (uint32_t *)lut_vector,
-        (uint32_t *)lut_vector_indexes, (uint32_t *)lwe_in,
-        (double2 *)bootstrapping_key, lwe_dimension, polynomial_size,
-        base_log, l_gadget, num_samples,
-        num_lut_vectors);
-    break;
   default:
-    break;
+    printf("Error: polynomial size should be one of 512, 1024, 2048");
```
Comment (mayeul-zama) [inline concrete-cuda/cuda/src/bootstrap_low_latency.cu:101]: Some context could be added

File: concrete-cuda/cuda/src/bootstrap_low_latency.cuh [hunk 5]
```diff
@@ -280,6 +280,15 @@ __host__ void host_bootstrap_low_latency(
     uint32_t l_gadget,
     uint32_t num_samples,
     uint32_t num_lut_vectors) {
+
+  assert(base_log <= 16);
```
Comment (mayeul-zama) [inline concrete-cuda/cuda/src/bootstrap_low_latency.cuh:284]: A print with some context could be added before aborting

File: concrete-cuda/cuda/src/bootstrap_low_latency.cuh [hunk 6]
```diff
@@ -280,6 +280,15 @@ __host__ void host_bootstrap_low_latency(
...
+  int number_of_sm = 0;
+  // The number of samples should be lower than SM/(4 * (k + 1) * l) (the 
+  // factor 4 being related to the occupancy of 50%). The only supported 
+  // value for k is 1, so k + 1 = 2 for now.
+  cudaDeviceGetAttribute(&number_of_sm, cudaDevAttrMultiProcessorCount, 0);
+  assert(num_samples <= number_of_sm / 4. / 2. / l_gadget);
```
Comment (mayeul-zama) [inline concrete-cuda/cuda/src/bootstrap_low_latency.cuh:290]: A print with some context could be added before aborting

File: concrete-cuda/cuda/src/bootstrap_wop.cu [hunk 7]
```diff
@@ -1,105 +1,113 @@
...
-        uint32_t max_shared_memory) {
+    void *v_stream,
+    void *glwe_out,
+    void *ggsw_in,
+    void *lut_vector,
+    uint32_t glwe_dimension,
+    uint32_t polynomial_size,
+    uint32_t base_log,
+    uint32_t l_gadget,
+    uint32_t r,
```
Comment (mayeul-zama) [inline concrete-cuda/cuda/src/bootstrap_wop.cu:12]: Ideally, put formatting changes in a separate commit to ease reviews :)

File: concrete-cuda/cuda/src/bootstrap_wop.cu [hunk 8]
```diff
@@ -1,105 +1,113 @@
...
+    break;
+  case 8192:
+    host_cmux_tree<uint32_t, int32_t, Degree<8192>>(
+        v_stream,
+        (uint32_t *) glwe_out, (uint32_t *) ggsw_in, (uint32_t *) lut_vector,
+        glwe_dimension, polynomial_size, base_log, l_gadget, r,
+        max_shared_memory);
+    break;
+  default:
+    printf("Error: polynomial_size should be one of 512, 1024, 2048, "
```
Comment (mayeul-zama) [inline concrete-cuda/cuda/src/bootstrap_wop.cu:52]: Some context could be added

File: concrete-cuda/cuda/src/bootstrap_wop.cu [hunk 9]
```diff
@@ -1,105 +1,113 @@
...
+    break;
+  case 8192:
+    host_cmux_tree<uint64_t, int64_t, Degree<8192>>(
+        v_stream,
+        (uint64_t *) glwe_out, (uint64_t *) ggsw_in,(uint64_t *) lut_vector,
+        glwe_dimension, polynomial_size, base_log, l_gadget, r,
+        max_shared_memory);
+    break;
+  default:
+    printf("Error: polynomial_size should be one of 512, 1024, 2048, "
```
Comment (mayeul-zama) [inline concrete-cuda/cuda/src/bootstrap_wop.cu:107]: Some context could be added

File: concrete-cuda/cuda/src/bootstrap_wop.cu [hunk 10]
```diff
@@ -157,7 +165,8 @@ void cuda_extract_bits_32(
         base_log_ksk, l_gadget_ksk, number_of_samples);
     break;
   default:
-    break;
+    printf("Error: lwe_dimension_before should be one of 512, 1024, 2048");
```
Comment (mayeul-zama) [inline concrete-cuda/cuda/src/bootstrap_wop.cu:168]: Some context could be added

File: concrete-cuda/cuda/src/bootstrap_wop.cuh [hunk 11]
```diff
@@ -298,6 +298,7 @@ void host_cmux_tree(
         uint32_t r,
         uint32_t max_shared_memory) {
 
+    assert(base_log <= 16);
```
Comment (mayeul-zama) [inline concrete-cuda/cuda/src/bootstrap_wop.cuh:301]: A print with some context could be added before aborting

File: concrete-cuda/cuda/src/bootstrap_wop.cu (hunk 7 above)
Comment (agnesLeroy) [reply concrete-cuda/cuda/src/bootstrap_wop.cu:12]: Oh sorry about it! Sure I'll be more careful next time.

**Comment by agnesLeroy** [comment]
@mayeul-zama I've removed the formatting part to make a dedicated PR for formatting + CI check that format is standard. I've added context to all asserts as well. :slightly_smiling_face:

**Comment by agnesLeroy** [comment]
@slab-ci gpu_test

File: concrete-cuda/cuda/src/bootstrap_wop.cuh [hunk 12]
```diff
@@ -298,9 +298,6 @@ void host_cmux_tree(
         uint32_t r,
         uint32_t max_shared_memory) {
 
-    assert(glwe_dimension == 1); // For larger k we will need to adjust the mask size
```
[the remark this answers is not in the document]
Comment (agnesLeroy) [reply concrete-cuda/cuda/src/bootstrap_wop.cuh:301]: It's moved in the .cu file instead for consistency with other asserts.

File: concrete-cuda/cuda/src/bootstrap_amortized.cu [hunk 13]
```diff
@@ -73,6 +73,11 @@ void cuda_bootstrap_amortized_lwe_ciphertext_vector_32(
     uint32_t lwe_idx,
     uint32_t max_shared_memory) {
 
+    assert(("Error (GPU amortized PBS): base log should be <= 16", base_log <= 16));
```
[the remark this answers is not in the document]
Comment (agnesLeroy) [reply concrete-cuda/cuda/src/bootstrap_amortized.cu:76]: I hesitated and did not do it because the argument wouldn't be used anywhere except the assert, but actually iy'll be good to have it when we extend support to k > 1 so I'll add it.

**Comment by agnesLeroy** [comment]
@slab-ci gpu_test

**Comment by agnesLeroy** [comment]
@slab-ci gpu_test
